File: python/seller.py
from __future__ import annotations

from typing import Any

import requests
from flask import Blueprint, jsonify, request

from storage import (
    create_shop_account,
    get_product,
    get_shop,
    get_shop_offers,
    list_products_short,
    list_shops,
    upsert_shop_offer,
)

seller_bp = Blueprint("seller", __name__)
RANKING_URL = "http://localhost:8090/rank/seller"


def rank_seller_offers(offers: list[dict[str, Any]]) -> list[dict[str, Any]]:
    try:
        response = requests.post(RANKING_URL, json={"offers": offers}, timeout=0.6)
        response.raise_for_status()
        return response.json().get("offers", offers)
    except Exception:
        return sorted(offers, key=lambda x: (x.get("delivery_days", 99), -x.get("stock", 0), x["price"]))


@seller_bp.post("/api/seller/register")
def register_seller() -> Any:
    payload = request.get_json(force=True)
    name = str(payload.get("name", "")).strip()
    city = str(payload.get("city", "")).strip()
    if not name or not city:
        return jsonify({"error": "name and city are required"}), 400

    return jsonify(create_shop_account(name, city)), 201


@seller_bp.get("/api/seller/list")
def seller_list() -> Any:
    return jsonify(list_shops())


@seller_bp.get("/api/seller/<int:seller_id>/offers")
def seller_offers(seller_id: int) -> Any:
    seller = get_shop(seller_id)
    if not seller:
        return jsonify({"error": "Seller not found"}), 404

    offers = rank_seller_offers(get_shop_offers(seller_id))
    return jsonify({"seller": seller, "offers": offers})


@seller_bp.post("/api/seller/<int:seller_id>/offers")
def upsert_offer(seller_id: int) -> Any:
    payload = request.get_json(force=True)
    required = ["product_id", "price", "stock", "delivery_days", "warranty_months"]
    if any(field not in payload for field in required):
        return jsonify({"error": f"required fields: {', '.join(required)}"}), 400

    if not get_shop(seller_id):
        return jsonify({"error": "Seller not found"}), 404

    product_id = int(payload["product_id"])
    if not get_product(product_id):
        return jsonify({"error": "Product not found"}), 404

    updated = upsert_shop_offer(seller_id, payload)
    return jsonify({"status": "ok", "updated": updated})


@seller_bp.get("/api/product/list")
def product_list() -> Any:
    return jsonify(list_products_short())

File: js/seller.js
const sellerSelect = document.getElementById("seller-select");
const productSelect = document.getElementById("product-select");
const sellerOffers = document.getElementById("seller-offers");
const searchOffer = document.getElementById("search-offer");

const sellerName = document.getElementById("seller-name");
const sellerCity = document.getElementById("seller-city");
const registerResult = document.getElementById("register-result");

const offerPrice = document.getElementById("offer-price");
const offerStock = document.getElementById("offer-stock");
const offerDelivery = document.getElementById("offer-delivery");
const offerWarranty = document.getElementById("offer-warranty");

let currentOffers = [];

function formatKzt(value) {
  return new Intl.NumberFormat("ru-RU").format(value) + " ₸";
}

function switchPage(name) {
  document.querySelectorAll(".page").forEach((page) => page.classList.remove("active"));
  document.querySelectorAll(".side-link").forEach((button) => button.classList.remove("active"));

  document.getElementById(`page-${name}`).classList.add("active");
  const activeButton = document.querySelector(`.side-link[data-page='${name}']`);
  if (activeButton) activeButton.classList.add("active");
}

document.querySelectorAll(".side-link").forEach((button) => {
  button.addEventListener("click", () => switchPage(button.dataset.page));
});
document.querySelector("[data-open-add]").addEventListener("click", () => switchPage("add"));
document.querySelector("[data-open-products]").addEventListener("click", () => switchPage("products"));

function renderOffers() {
  const query = searchOffer.value.trim().toLowerCase();
  const filtered = currentOffers.filter((offer) => offer.product_title.toLowerCase().includes(query));

  sellerOffers.innerHTML = filtered.length
    ? filtered.map((offer) => `
      <tr>
        <td>${offer.product_title}</td>
        <td>${formatKzt(offer.price)}</td>
        <td>RP1</td>
        <td>${offer.stock}</td>
      </tr>
    `).join("")
    : `<tr><td colspan="4" class="muted">Нет подходящих товаров.</td></tr>`;
}

async function loadLists() {
  const sellers = await (await fetch("/api/seller/list")).json();
  const products = await (await fetch("/api/product/list")).json();

  sellerSelect.innerHTML = sellers.map((seller) => `<option value="${seller.id}">${seller.name} (${seller.city})</option>`).join("");
  productSelect.innerHTML = products.map((product) => `<option value="${product.id}">${product.title}</option>`).join("");

  if (sellers.length) {
    await loadSellerOffers(sellers[0].id);
  }
}

async function loadSellerOffers(sellerId) {
  const response = await fetch(`/api/seller/${sellerId}/offers`);
  const payload = await response.json();

  if (!response.ok) {
    currentOffers = [];
    sellerOffers.innerHTML = `<tr><td colspan="4" class="muted">Ошибка: ${payload.error}</td></tr>`;
    return;
  }

  currentOffers = payload.offers;
  renderOffers();
}

sellerSelect.addEventListener("change", () => loadSellerOffers(sellerSelect.value));
searchOffer.addEventListener("input", renderOffers);

document.getElementById("register-seller").addEventListener("click", async () => {
  const response = await fetch("/api/seller/register", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ name: sellerName.value.trim(), city: sellerCity.value.trim() }),
  });

  const payload = await response.json();
  registerResult.textContent = response.ok ? `Создан продавец: ID ${payload.id}` : `Ошибка: ${payload.error}`;

  if (response.ok) {
    await loadLists();
  }
});

document.getElementById("save-offer").addEventListener("click", async () => {
  const sellerId = sellerSelect.value;
  const body = {
    product_id: Number(productSelect.value),
    price: Number(offerPrice.value),
    stock: Number(offerStock.value),
    delivery_days: Number(offerDelivery.value),
    warranty_months: Number(offerWarranty.value),
  };

  const response = await fetch(`/api/seller/${sellerId}/offers`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  });

  const payload = await response.json();
  if (!response.ok) {
    alert(payload.error);
    return;
  }

  await loadSellerOffers(sellerId);
  alert(payload.updated ? "Оффер обновлен" : "Оффер добавлен");
});

loadLists();
